cmd/krel/cmd/set_release_version.go

```diff
@@ -0,0 +1,155 @@
+	Long: `krel set-release-version
+
+This subcommand can be used to set the release version from a bash command by
+sourcing it's output. It's mainly indented to be used from anago, which means
+the command might be removed in future releases again if anago is end of life.
+`,
+	SilenceUsage:  true,
+	SilenceErrors: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		res, err := runSetReleaseVersion(setReleaseVersionOpts)
+		if err != nil {
+			return err
+		}
+		fmt.Print(res)
+		return nil
+	},
+}
+
+type setReleaseVersionOptions struct {
+	releaseType  string
+	buildVersion string
+	branch       string
+	parentBranch string
+}
+
+var setReleaseVersionOpts = &setReleaseVersionOptions{}
+
+func init() {
+	setReleaseVersionCmd.PersistentFlags().StringVar(
+		&setReleaseVersionOpts.releaseType,
+		"release-type",
+		"",
+		fmt.Sprintf("release type, must be one of: '%s'",
+			strings.Join([]string{
+				release.ReleaseTypeAlpha,
+				release.ReleaseTypeBeta,
+				release.ReleaseTypeRC,
+				release.ReleaseTypeOfficial,
+			}, "', '"),
+		),
+	)
+
+	setReleaseVersionCmd.PersistentFlags().StringVar(
+		&setReleaseVersionOpts.buildVersion,
+		"build-version",
+		"",
+		"jenkins build version to be used",
+	)
+
+	setReleaseVersionCmd.PersistentFlags().StringVar(
+		&setReleaseVersionOpts.branch,
+		"branch",
+		"",
+		"branch for which the version should be calculated",
+	)
+
+	setReleaseVersionCmd.PersistentFlags().StringVar(
+		&setReleaseVersionOpts.parentBranch,
+		"parent-branch",
+		"",
+		"the parent branch for the target branch",
+	)
+
+	for _, f := range []string{
+		"release-type",
+		"build-version",
+		"branch",
+		"parent-branch",
+	} {
+		if err := setReleaseVersionCmd.MarkPersistentFlagRequired(f); err != nil {
+			logrus.Fatalf("Unable to set %q flag as required: %v", f, err)
+		}
+	}
+
+	rootCmd.AddCommand(setReleaseVersionCmd)
+}
+
+func runSetReleaseVersion(opts *setReleaseVersionOptions) (string, error) {
+	releaseVersion, err := release.SetReleaseVersion(
+		opts.releaseType, opts.buildVersion, opts.branch, opts.parentBranch,
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "set release version")
+	}
+
+	res := strings.Builder{}
+	res.WriteString("declare -Ag RELEASE_VERSION\n")
+	res.WriteString("declare -ag ORDERED_RELEASE_KEYS\n")
+
+	addKey := func(v string) {
+		res.WriteString(fmt.Sprintf("ORDERED_RELEASE_KEYS+=(%q)\n", v))
+	}
+
+	addVersion := func(k, v string) {
+		res.WriteString(fmt.Sprintf("RELEASE_VERSION[%s]=%q\n", k, v))
+	}
+
+	if releaseVersion.Official() != "" {
+		const v = "official"
+		addVersion(v, releaseVersion.Official())
+		addKey(v)
+	}
+	if releaseVersion.RC() != "" {
+		const v = "rc"
+		addVersion(v, releaseVersion.RC())
+		addKey(v)
+	}
+	if releaseVersion.Beta() != "" {
+		const v = "beta"
+		addVersion(v, releaseVersion.Beta())
+		addKey(v)
+	}
+	if releaseVersion.Alpha() != "" {
+		const v = "alpha"
+		addVersion(v, releaseVersion.Alpha())
+		addKey(v)
+	}
```

Comment:
Can this be a `switch`?

Reply:
Hm, since we have different methods to access here I don't think that a switch statement would help us to make it more straight forward to read. How would it look like? :)

Reply:
I think switch is not needed here, it is not one variable with different values.

Reply:
The main intention for the type with different assessors was to not rely on a map. Maybe this decision makes it more complicated to use the type in the future. I'll revisit and check how we can improve it.